DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+
+namespace DataStructures.Lists 
+{
+    public class CircularBuffer<T> 
+    {
+        /// <summary>
+        /// Initializes a circular buffer with initial length of 10
+        /// </summary>
+        public CircularBuffer(bool canOverride=true) : this(_defaultBufferLength, canOverride) 
+        {
+        }
+
+        /// <summary>
+        /// Initializes a circular buffer with given length
+        /// </summary>
+        /// <param name="length">The length of the buffer</param>
+        public CircularBuffer(int length, bool canOverride=true) 
+        {
+            if (length < 1) 
+            {
+                throw new ArgumentOutOfRangeException("length can not be zero or negative");
+            }
+            _circularBuffer = new T[length + 1];
+            _end = 0;
+            _start = 0;
+            CanOverride = canOverride;
+        }
+
+        /// <summary>
+        /// Writes value to the back of the buffer
+        /// </summary>
+        /// <param name="value">value to be added to the buffer</param>
+        public void Write(T value) 
+        {
+            if (CanOverride) 
+            {
+                InsertData(value);
+            }
+            else 
+            {
+                DontOverrides(value);
+            }
+        }
+
+        // Inserts data into the buffer when it is not filled up
+        private void DontOverrides(T value) 
+        {
+            if (IsFilledUp) 
+            {
+                throw new CircularBufferFullException($"Circular Buffer is filled up. {value} can not be inserted");
+            }
+            InsertData(value);
+        }
+
+        // Inserts data into the buffer without checking if it is full
+        private void InsertData(T value) 
```

Comment:
Please use `thisCase` for private methods and `ThatCase` for public methods. It will help people visually understand the difference. Also, it would make more sense if you renamed the method to `innerInsert` than `InsertData`, as the `Data` word here is not needed.

Reply:
Okay. 
Done